core/jvm/src/SegmentPool.kt

```diff
@@ -25,34 +25,81 @@ import kotlinx.io.SegmentPool.LOCK
 import kotlinx.io.SegmentPool.MAX_SIZE
 import kotlinx.io.SegmentPool.recycle
 import kotlinx.io.SegmentPool.take
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater
 import java.util.concurrent.atomic.AtomicReference
 
+/**
+ * Precise [SegmentCopyTracker] implementation tracking a number of shared segment copies.
+ * Every [addCopy] call increments the counter, every [removeCopyIfShared] decrements it.
+ *
+ * After calling [removeCopyIfShared] the same number of time [addCopy] was called, tracker returns to the unshared state.
+ *
+ * The class is internal for testing only.
+ */
+internal class RefCountingCopyTracker : SegmentCopyTracker() {
+    companion object {
+        @JvmStatic
+        private val fieldUpdater = AtomicIntegerFieldUpdater.newUpdater(RefCountingCopyTracker::class.java, "refs")
+    }
+
+    @Volatile
+    private var refs: Int = 0
+
+    override fun addCopy() {
+        fieldUpdater.incrementAndGet(this)
+    }
+
+    override fun removeCopyIfShared(): Boolean {
+        while (true) {
+            val value = refs
+            check(value >= 0) { "Shared copies count is negative: $value" }
+            if (fieldUpdater.compareAndSet(this, value, (value - 1).coerceAtLeast(0))) {
```

Comment:
When returning a non-sharing segment into a pool, its `refs` value (if a tracker was allocated at all) will be `0` and it would be incorrect to decrement it further.

Reply:
Reworked this part after Sergey's suggestions.